### PR TITLE
Add grouping tests

### PR DESCRIPTION
**Pull request by DougBurke**
# Summary

Add a number of tests to check out corner case behavior, primarily for grouping PHA data but there's also a number of routines from the UI code that have missed some tests.

# Details

There are a number of PRs that require this PR to be merged:

- #1486 (and a bunch of follow-up PRs)
- #1664 
- #1665 
- #1666

Tests have been found to test current behavior, and to track new bugs found from this work (with the intention being that we fix these in later PRs). Some of these tests have been found during PR development, by seeing what code changes that were made had no existing tests.

Issues with new (i.e. were found doing this work) or extended coverage:

- #1160 
- #1641
- #1642 
- #1646 
- #1659

### CODE REVIEW

**Comment by DougBurke** [comment]
@hamogu - I've split up #1638 into a number of separate PRs. They all rely on this PR, which just adds tests. Some of the tests just ensure coverage for code (varies from "check we error out in this condition" to "this particular show_xxx routine had no test"), and some show bugs we are going to fix. You can see that both the `utils.py` files have more than 3% increase in coverage (and they are both now at at least 95% coverage); now coverage isn't everything, but it is useful to have!

I've manually added a fix from #1663 which will need to be rebased (but it's a simple fix/change and unrelated to this code).

`sherpa/astro/ui/tests/test_astro_session.py` is interesting as I've added a pytest fixture `skip_if_no_io` which is a bit like the existing `requires_io` decorator but more specialized (we only skip when using the `sherpa.astro.ui.utlls.Session` class but run it with `sherpa.ui.utils.Session`). Previously I used `requires_io` for these tests, so this runs a few more tests (there are routines in `sherpa.ui.utils.Session` which have a fall-through when there's no I/O backend that do not behave the same way in `sherpa.astro.ui.utils.Session`) - see #1668.

File: sherpa/astro/ui/tests/test_astro_ui.py [hunk 1]
```diff
@@ -513,6 +703,51 @@ def test_psf_model2d(model, clean_astro_ui):
             numpy.array([108, 130])).all()
 
 
+def test_dataspace2d_default(clean_astro_ui):
+    """What happens when we calll dataspace2d: default"""
```
Comment (hamogu) [inline sherpa/astro/ui/tests/test_astro_ui.py:707]: calll -> call (same in the next few examples)

File: sherpa/astro/tests/test_astro_data2.py [hunk 2]
```diff
@@ -1575,11 +1713,29 @@ def test_pha_remove_grouping(make_test_pha):
 
     # Can we remove the grouping column?
     pha.grouping = None
-    # This thinks that pha.grouped is still set
-    assert not pha.grouped
+
+    # Check the get_dep ehavior before grouped as this is currently
```
Comment (hamogu) [inline sherpa/astro/tests/test_astro_data2.py:1717]: ehavior -> behavior

File: sherpa/astro/tests/test_astro_data2.py [hunk 3]
```diff
@@ -3773,11 +3957,11 @@ def test_pha_subtract_bkg_filter_false():
     data.set_background(bkg)
     data.subtract()
 
-    bgot = bkg.get_staterror(filter=False, staterrfunc=lambda x: np.sqrt(x))
+    bgot = bkg.get_staterror(filter=False, staterrfunc=np.sqrt)
```
Comment (hamogu) [inline sherpa/astro/tests/test_astro_data2.py:3960]: Should we test that the old lambda case also works? In this case, `lambda x: np.sqrt(x)` is of course identical to `np.sqrt`, but maybe a user wants something different like `lambda x: np.sqrt(x + 1)`. Not necessarily in scope for this test, but I wonder if that was the motivation behind using the lambda here originally.

**Review by hamogu — CHANGES_REQUESTED** [review]
This is somewhat of an odd review. Since it's all added tests and CI passes, it's almost correct by definition. All the tests seem to make sense. Are all strictly required? Should have have more tests? I think both questions are out of scope for this PR. The PR adds tests and if I wanted more, I can always add more later. And what's the harm if we have too mant tests? Well, then we waste a few mW of electricity since this all runs very fast. So, my only comments are typos in the docstrings; but I'll select "request changes", because formally that's something where I, well, request changes, even if it does not matter in the slightest.

Note that the last commit in this PR is #1663, so that one should probably be merged.

File: .github/scripts/setup_conda.sh [hunk 4]
```diff
@@ -18,16 +18,15 @@ else
...
         #  ldd $CONDA_PREFIX/plugins/platforms/libqxcb.so | grep "not found"
         sudo apt-get install -q libxcb-icccm4 libxcb-image0 libxcb-keysyms1 libxcb-render-util0 libxcb-icccm4 libxcb-image0 libxcb-keysyms1 libxcb-render-util0
    fi
 fi
 
 # Download and install conda
 wget https://repo.anaconda.com/miniconda/Miniconda3-latest-${miniconda_os}-x86_64.sh -O miniconda.sh
-chmod +x miniconda.sh
-./miniconda.sh -b -p $miniconda_loc
+bash miniconda.sh -b -p $miniconda_loc
```
[the remark this answers is not in the document]
Comment (DougBurke) [reply .github/scripts/setup_conda.sh:29]: Yeah - this commit was just needed to let the tests run - we need it addressed outside of this PR first.

File: sherpa/astro/ui/tests/test_astro_ui_unit.py [hunk 5]
```diff
@@ -1741,6 +1922,30 @@ def test_get_axes_datapha_rsp():
     assert ax[1] == pytest.approx([0.2, 0.4, 0.8])
 
 
+def test_get_axes_datapha_rsp_bkg(clean_astro_ui):
+    """Let's have a RMF and ARF for fun (but only for source)
+
+    It is not clear whether we expect the respnose to be automatically
+    applied to the background here, so treat this as a regression test
+    (because, at present, we do not have units=energy for the
+    background(.
```
[the remark this answers is not in the document]
Comment (DougBurke) [reply sherpa/astro/ui/tests/test_astro_ui_unit.py:1931]: Ha - you missed the `respnose` typo a few lines earlier ;-)

File: sherpa/astro/tests/test_astro_data2.py (hunk 3 above)
Comment (DougBurke) [reply sherpa/astro/tests/test_astro_data2.py:3960]: This was a pylint-inspired change. Given the mechanics of function application I don't see the need for something clever, since "lambda x: np.sqrt(x)" and "np.sqrt" are essentially the same. I don't see the need to add something extra.

File: sherpa/astro/ui/tests/test_astro_ui_unit.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (hamogu) [reply sherpa/astro/ui/tests/test_astro_ui_unit.py:1931]: Did you put that in just to test me? I'm running a spell checker to spell-check any comments in my editor now... but not on PR review.

File: sherpa/astro/utils/tests/test_astro_utils_unit.py [hunk 6]
```diff
@@ -581,3 +582,90 @@ def test_calc_data_sum2d_filtered_2d(data_class):
...
+    assert len(out) == 0
+
+
+@pytest.mark.parametrize("val", [None, 0, 1, "x"])  # mask is converted to bool so a string/None are accepted
+@pytest.mark.parametrize("group", [[1], [1, -1], [1] * 10, []])
+def test_expand_grouped_mask_mask_is_scalar(val, group):
+    """What happens if mask is a scalar.
+
+    This is an edge case. It should probably error out but at the
+    moment it does not.
```
[the remark this answers is not in the document]
Comment (DougBurke) [reply sherpa/astro/utils/tests/test_astro_utils_unit.py:611]: This is changed in #1486 - the "Add validation to expand_grouped_mask" commit (I won't link to it as it'll just go out of date)

**Comment by DougBurke** [comment]
I've rebased this to address #1669 being merged.

File: .github/scripts/setup_conda.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (DougBurke) [reply .github/scripts/setup_conda.sh]: This has now been removed as #1669 has been merged

**Review by hamogu — APPROVED** [review]
